dataset_generator/identify_partial_borrowers.py

```diff
@@ -56,18 +64,114 @@ def identify_partial_borrowers():
             # if there are no book events within this subscription,
             # add it to the list of partial borrower dates
             if sub_book_events.empty:
+                # print('\n%s' % member_id)
+                # print(f"{sub.start_date} - {sub.end_date} ({sub.event_type}, {sub.subscription_volumes} volumes)")
                 partial_borrowers.append(
                     {
                         "member_id": member_id,
                         "subscription_start": sub.start_date,
                         "subscription_end": sub.end_date,
+                        "subscription_type": sub.event_type,
+                        "subscription_volumes": sub.subscription_volumes,
                         "known_borrows": len(member_book_events.index),
                     }
                 )
 
     # load into dataframe and save as csv for use elsewhere
-    pd.DataFrame(data=partial_borrowers).to_csv(PARTIAL_BORROWERS_CSV, index=False)
+    partial_borrows = pd.DataFrame(data=partial_borrowers)
+    partial_borrows.to_csv(PARTIAL_BORROWERS_CSV, index=False)
+
+    # now collapse sequential-ish subscription periods for each member
+    collapse_partial_borrowers(partial_borrows)
+
+
+def collapse_partial_borrowers(partial_borrows):
+    # collapse sequential periods into single time periods
+    collapsed_ranges = []
+    # convert to datetime for comparison
+    partial_borrows["start_date"] = pd.to_datetime(
+        partial_borrows.subscription_start, format="%Y-%m-%d", errors="coerce"
+    )
+    partial_borrows["end_date"] = pd.to_datetime(
+        partial_borrows.subscription_end, format="%Y-%m-%d", errors="coerce"
+    )
+
+    # allowed gap between sequential periods
+    # ?? what amount of time is reasonable here?
+    allowed_gap = pd.Timedelta(days=90)
+
+    for member_id in partial_borrows.member_id.unique():
+        range_start = None
+        range_end = None
+
+        # get the previously identified subscriptions with no borrowing
+        member_subs = partial_borrows[partial_borrows.member_id == member_id]
+        for sub in member_subs.itertuples():
```

Comment:
I'm so excited about `itertuples` I didn't know about this!

Reply:
so useful, I use it all the time